jerry-core/ecma/base/ecma-helpers-conversion.c

```diff
@@ -457,6 +484,57 @@ ecma_utf8_string_to_number (const lit_utf8_byte_t *str_p, /**< utf-8 string */
     return ecma_number_make_infinity (sign);
   }
 
+  if (!is_infinity_check)
```

Comment:
I'm almost 100% percent sure now that having an "infinity check" argument and a two step recursion is not the way to go. I did a small experiment and tried to do a manual constant propagation by replacing `is_infinity_check` with `true` everywhere and then removing all that code that is under an `if (false)` (or equivalent), to see what code executed in the second step looks like. And it eliminates a huge portion of the coode. (And the same thing happens if I do the same with propagating `false` around to see how the first step looks like.) All this hints at (at least) two independent code paths being badly intermixed.

And a lot of code in this block is a repetition of work that is also handled later (parsing of exponent).